lib/lib-search/src/main/java/org/sagebionetworks/search/CloudSearchClientProvider.java

```diff
@@ -0,0 +1,89 @@
+package org.sagebionetworks.search;
+
+import com.amazonaws.services.cloudsearchdomain.AmazonCloudSearchDomainClient;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.sagebionetworks.repo.web.TemporarilyUnavailableException;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.util.Random;
+
+/**
+ * This class is responsible for initializing the setup of AWS CloudSearch
+ * and providing a configured CloudSearch client once setup has completed.
+ */
+public class CloudSearchClientProvider {
+	static private Logger log = LogManager.getLogger(CloudSearchClientProvider.class);
+
+	@Autowired
+	SearchDomainSetup searchDomainSetup;
+
+	@Autowired
+	AmazonCloudSearchDomainClient awsCloudSearchDomainClient;
+
+	private boolean isSearchEnabled;
+
+	private static CloudsSearchDomainClientAdapter singletonWrapper;
+
+	public CloudsSearchDomainClientAdapter getCloudSearchClient(){
+		if(!isSearchEnabled()){
+			throw new SearchDisabledException();
+		}
+
+		if (singletonWrapper != null){
+			return singletonWrapper;
+		}else{
+			if(searchDomainSetup.postInitialize()) {
+				awsCloudSearchDomainClient.setEndpoint(searchDomainSetup.getDomainSearchEndpoint());
+				singletonWrapper = new CloudsSearchDomainClientAdapter(awsCloudSearchDomainClient);
+				return singletonWrapper;
+			} else{
+				log.warn("CloudSearch is not finished initializing");
+				throw new TemporarilyUnavailableException("Search has not yet been initialized. Please try again later!");
+			}
+		}
+	}
+
+	/**
+	 * Injected via Spring
+	 * @param isSearchEnabled
+	 */
+	public void setSearchEnabled(boolean isSearchEnabled) {
+		this.isSearchEnabled = isSearchEnabled;
+	}
+
+	public boolean isSearchEnabled() {
+		return isSearchEnabled;
+	}
+
+	/**
+	 * The initialization of a search index can take hours the first time it is run.
+	 * While the search index is initializing we do not want to block the startup of the rest
+	 * of the application.  Therefore, this initialization worker is executed on a separate
+	 * thread.
+	 */
+	public void initialize() {
```

Comment:
If initialize only occurs when a search is run (not on startup), the stack builder can trigger a search to trigger the initialize.  Then we no longer need this "hacky" random sleep.